### PR TITLE
[tiktok] fix extraction

### LINKED ISSUE #20809: Can't download from Tiktok
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2019.04.24**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/ytdl-org/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/ytdl-org/youtube-dl#faq) and [BUGS](https://github.com/ytdl-org/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
youtube-dl.exe --verbose https://www.tiktok.com/share/video/6682711605567884550
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'https://www.tiktok.com/share/video/6682711605567884550']
[debug] Encodings: locale cp1251, fs mbcs, out cp866, pref cp1251
[debug] youtube-dl version 2019.04.24
[debug] Python version 3.4.4 (CPython) - Windows-10-10.0.17763
[debug] exe versions: ffmpeg 3.4.1, ffprobe N-68994-g4df01d5, phantomjs 2.1.1
[debug] Proxy map: {}
[TikTok] 6682711605567884550: Downloading webpage
ERROR: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp1l3rbf78\build\youtube_dl\YoutubeDL.py", line 796, in extract_info
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp1l3rbf78\build\youtube_dl\extractor\common.py", line 529, in extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp1l3rbf78\build\youtube_dl\extractor\tiktok.py", line 101, in _real_extract
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\ytdl-org\tmp1l3rbf78\build\youtube_dl\extractor\common.py", line 1004, in _search_regex
youtube_dl.utils.RegexNotFoundError: Unable to extract data; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

---

I can't download any TikTok video and supported sites lists it as supported site.

### PR DESCRIPTION
**Pull request by skyme5**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
Extractor did not work. 

- Extract more metadata
- Can only extract latest 9 videos uploaded by user
- Fixes #20809, #22850 and #23264

### CODE REVIEW

**Comment by dstftw** [comment]
Read coding conventions.

**Comment by Qwerty-Space** [comment]
Is this in the latest version?

**Comment by BryanYin** [comment]
Hi, when will this expected to be merged? 
Thanks!

**Comment by Qwerty-Space** [comment]
@dstftw

**Comment by unexpectedShr3k** [comment]
will this be merged anytime soon?

**Comment by skyme5** [comment]
@dstftw When this can be merged ?

**Comment by karottenbaum** [comment]
With the patch applied, a `RegexNotFoundError` for `json_string` results for me.
<details>
<summary>Debug information</summary>

```
[debug] System config: []
[debug] User config: ['--all-subs', '--fragment-retries', '10', '--no-mark-watched', '--abort-on-unavailable-fragment', '--console-title', '--write-all-thumbnails']
[debug] Custom config: []
[debug] Command-line args: ['https://www.tiktok.com/@ravirosss123/video/6773432341571456261', '-v']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.10.22
[debug] Git HEAD: 4b6d03ed8
[debug] Python version 3.6.9 (CPython) - Linux-5.0.0-37-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 4.0.2-1, ffprobe 4.0.2-1, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[TikTok] 6773432341571456261: Downloading video webpage
ERROR: Unable to extract json_string; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/home/chris/github/youtube-dl/youtube_dl/YoutubeDL.py", line 796, in extract_info
    ie_result = ie.extract(url)
  File "/home/chris/github/youtube-dl/youtube_dl/extractor/common.py", line 530, in extract
    ie_result = self._real_extract(url)
  File "/home/chris/github/youtube-dl/youtube_dl/extractor/tiktok.py", line 243, in _real_extract
    webpage, 'json_string', group='json_string')
  File "/home/chris/github/youtube-dl/youtube_dl/extractor/common.py", line 1005, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name)
youtube_dl.utils.RegexNotFoundError: Unable to extract json_string
```
</details>

**Comment by skyme5** [comment]
> With the patch applied, a `RegexNotFoundError` for `json_string` results for me.
> Debug information

@karottenbaum Fixed the issue in the latest commit.

**Comment by thinkpad4** [comment]
Is there a way to sideload the fixed tiktok.py file into youtube-dl until it is pushed out fully?

**Comment by skyme5** [comment]
> Is there a way to sideload the fixed tiktok.py file into youtube-dl until it is pushed out fully?

@thinkpad4 you can clone this repository and run from it.
```bash
git clone https://github.com/skyme5/youtube-dl.git
cd youtube-dl/youtube_dl
python __main__.py url
```

**Comment by Qwerty-Space** [comment]
@dstftw all checks have passed, can you merge?

File: youtube_dl/extractor/tiktok.py [hunk 1]
```diff
@@ -1,138 +1,252 @@
...
         video_id = self._match_id(url)
-        webpage = self._download_webpage(
-            'https://m.tiktok.com/v/%s.html' % video_id, video_id)
-        data = self._parse_json(self._search_regex(
-            r'\bdata\s*=\s*({.+?})\s*;', webpage, 'data'), video_id)
-        return self._extract_aweme(data)
 
+        headers = {
+            'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:72.0) Gecko/20100101 Firefox/72.0',
+            'Referer': url
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: Referring `url` from `url` looks like nonsense. Provide rationale.

File: youtube_dl/extractor/tiktok.py [hunk 2]
```diff
@@ -1,138 +1,252 @@
...
+        }
+        webpage = self._download_webpage(url, video_id, headers=headers, note='Downloading video webpage')
+        json_string = self._search_regex(
+            r'id=\"__NEXT_DATA__\"\s+type=\"application\/json\"\s*[^>]+>\s*(?P<json_string>[^<]+)',
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: `_search_json_ld`.

File: youtube_dl/extractor/tiktok.py [hunk 3]
```diff
@@ -1,138 +1,252 @@
...
-                            (?:
-                                (?:m\.)?tiktok\.com/v|
-                                (?:www\.)?tiktok\.com/share/video
-                            )
-                            /(?P<id>\d+)
-                    '''
+                 https?://
+                 (?:
+                    (?:www|m)\.
+                    (?:tiktok.com)\/
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: Useless group.

File: youtube_dl/extractor/tiktok.py [hunk 4]
```diff
@@ -1,138 +1,252 @@
...
+                    (@(?P<username>[\w\.]+))?
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: Don't capture groups you don't use. Unused captured group.

File: youtube_dl/extractor/tiktok.py [hunk 5]
```diff
@@ -1,138 +1,252 @@
...
+                    (?:v|video|embed|trending)?(?:\/)?
+                    (?:video)?(?:\/)?
+                    (?:\?shareId=)?
+                 )
+                 (?P<id>[\d]{6,})
+                 (?:\.html)?
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: This is no longer actual.

File: youtube_dl/extractor/tiktok.py [hunk 6]
```diff
@@ -1,138 +1,252 @@
...
+                 https?://
+                 (?:
+                    (?:www|m)\.
+                    (?:tiktok.com)\/
+                    (@(?P<username>[\w\.]+))?
+                    (?:v|video|embed|trending)?(?:\/)?
+                    (?:video)?(?:\/)?
+                    (?:\?shareId=)?
+                 )
+                 (?P<id>[\d]{6,})
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: No brackets needed.

File: youtube_dl/extractor/tiktok.py [hunk 7]
```diff
@@ -1,138 +1,252 @@
...
-                            /(?P<id>\d+)
-                    '''
+                 https?://
+                 (?:
+                    (?:www|m)\.
+                    (?:tiktok.com)\/
+                    (@(?P<username>[\w\.]+))?
+                    (?:v|video|embed|trending)?(?:\/)?
+                    (?:video)?(?:\/)?
+                    (?:\?shareId=)?
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: This is total mess. Should not match https://www.tiktok.com/@leenabhushanvideo6748451240264420610, https://www.tiktok.com/embed/video/6567659045795758085 and so on.

File: youtube_dl/extractor/tiktok.py [hunk 8]
```diff
@@ -1,138 +1,252 @@
...
 )
 
 
 class TikTokBaseIE(InfoExtractor):
-    def _extract_aweme(self, data):
-        video = data['video']
-        description = str_or_none(try_get(data, lambda x: x['desc']))
-        width = int_or_none(try_get(data, lambda x: video['width']))
-        height = int_or_none(try_get(data, lambda x: video['height']))
+    def _video_info(self, video_info):
```
Comment (dstftw) [inline youtube_dl/extractor/tiktok.py]: Inline everything used only once.

**Pull request closed without merging**